### DISCUSSION TITLE
ESP32  Schematic of Waveshare General Driver for Robot

### DISCUSSION THREAD

**Discussion by hippo5329** [hippo5329/linorobot2_hardware#12] [Announcements]
The ESP32 Schematic of Waveshare General Driver for Robot can be found at their Wiki.

[Wiki](https://www.waveshare.com/wiki/General_Driver_for_Robots) 
[Schematic](https://files.waveshare.com/upload/3/37/General_Driver_for_Robots.pdf).

**Comment by otomoov** [comment]
Hello and good day to you. I have a question. Is the Raspberry Pi 40 pins on Waveshare General Driver for Robot connected to the ESP32 38 pins?

**Comment by hippo5329** [comment]
esp32 gpio 32 i2c sda is connected to rpi pin 3 sda gpio 2
esp32 gpio 33 i2c scl is connected to rpi pin 5 scl gpio 3

**Comment by otomoov** [comment]
Thank you for the reply.

![IMG_20240916_073903](https://github.com/user-attachments/assets/b463aa30-ac00-45d6-b367-a45908f9851c)

I did continuity test between the raspberry pi style 40pin header and the esp32 pins. No continuity except for SDA, SCL, TX, RX, 5V,  GND.

I did esp32 blink test on all io pins. No blinking on the raspberry pi breakout board.

![Screenshot_2024-09-16-07-48-37-237_com google android apps docs-edit](https://github.com/user-attachments/assets/6b5948c7-31e0-445b-a0fe-4155ca473ec8)
Based on this drawing I see no gpio connection other than SDA, SCL, TX, RX.

I was hoping the 40pin header could be use for button signal input and light signal output. Is it possible to do it with Waveshare General Driver for Robot ?

**Comment by hippo5329** [comment]
The waveshare general robot driver board is very limited in extra io pins. I would suggest add another esp32/pico to control the led and keypad. They are cheap and easy to use.

gendrv linorobot2_hardware micro-ros-- usb serial -- rpi
esp32 keypad rgb led micro-ros --usb serial -- rpi

A single micro-ros agent can control multiple serial ports.

https://micro-xrce-dds.docs.eprosima.com/en/latest/agent.html

Interfaces,

keypad msg -- int32 publisher
rgb led msg -- int32 subscriber

keypad

https://github.com/Chris--A/Keypad
https://www.circuitbasics.com/how-to-set-up-a-keypad-on-an-arduino/

rgb led

https://www.instructables.com/Driving-RGB-LED-strips-off-an-Arduino/
control the leds with analogwrite to drive the MOSFET

or you can use p9813 led driver module and fastled driver

https://github.com/FastLED/FastLED

![Screenshot from 2024-09-16 10-04-59](https://github.com/user-attachments/assets/3bcc4346-ed56-4c81-81b7-38123dfa8125)



I am traveling and do not have the hardware to test until next month.

**Comment by hippo5329** [comment]
micro-ros multiserial usage,

ros2 run micro_ros_agent micro_ros_agent multiserial --devs "/dev/ttyUSB0 /dev/ttyUSB1" --baudrate 921600

You may need to use by-path dev names.
